### DISCUSSION TITLE
Poor performance when trying to select span name

### DISCUSSION THREAD

**Discussion by RonanQuigley** [grafana/tempo#3617] [Q&A] [answered]
If I have a large number of span names to retrieve, the UI will struggle to serve the span name dropdown quickly. It will take seconds to return the span names or just fail.  

![slow](https://github.com/grafana/tempo/assets/33596237/acb617cc-2370-40fb-8c61-c91f1b54b1f5)

I don't see any obvious issues in the tempo setup that could cause this. Here's the CPU metrics.

![image](https://github.com/grafana/tempo/assets/33596237/b3160e0c-ead8-457a-8415-30fa5533eb7e)

Is there anything I can tweak to improve performance here e.g. limit the returned number of span names? The limit input field makes no difference

**Comment by joe-elliott** [comment]
What version of Tempo and Grafana are you running?
Can you use the network tab to see what endpoint Grafana is hitting? How long is taking the endpoint returning?

> If I have a large number of span names to retrieve

What is a "large number"?

**Reply by RonanQuigley** [reply]
Hey, thanks for responding. 

Sure, here's the details: 

Endpoint: `api/search/tag/name/values` 
Grafana Version: 9.5.5 

Tempo:
Chart: tempo-distributed-1.8.2
App Version: 2.3.1

There are 9918 span names being returned. 

I did some further investigation into the problem after my initial post. The issue is that Grafana's UI will struggle to load that many elements into a dropdown quickly. The CPU usage maxes to 100% and the UI thread gets blocked:
  
![image](https://github.com/grafana/tempo/assets/33596237/a29bc223-79fc-4fc2-92fe-6caa17bf72d6).

Furthermore, if I select a specific service, it returns span names that are unrelated to that service, so you can't workaround this problem by filtering by service name first. 

Shall I try update Grafana to the latest version to see it fixes this? I was trying to find related issues/PRs that looked into this on the main grafana repo but I couldn't find any obvious results.

**Reply by joe-elliott** [reply]
> Shall I try update Grafana to the latest version to see it fixes this?

Yup. I think that's your best bet. Current Grafana has made a lot of performance improvements with regard to loading longer select lists.

> Furthermore, if I select a specific service, it returns span names that are unrelated to that service, so you can't workaround this problem by filtering by service name first.

Also, Tempo has added a `v2` of the attribute values endpoint. This allows filtering by other fields. So in current Grafana selecting a service first will nicely scope down your span names.

**Reply by RonanQuigley** [reply]
Hey, we updated to Grafana v11.0.0 (b12ad73c52) and are still seeing issues. It is now using the v2 endpoint, but the UI will stall. It's therefore hard for me to tell if the span names are indeed being scoped down.

Shall I raise this over in the main Grafana repo? There needs to be a cap placed on how many span names are listed.  This is what we could see for tags as an example.

![image](https://github.com/grafana/tempo/assets/33596237/98924f10-e818-4750-8e91-5135bf209eda).

**Reply by joe-elliott** [reply]
Yeah, I think Grafana is trying to render all 78K+ names at once. Filing an issue in the Grafana repo is your best bet.

Can you share how long it is taking Tempo to return that information?

**Reply by RonanQuigley** [reply, marked as answer]
I've submitted a PR to fix this 

https://github.com/grafana/grafana/pull/88569
